File: 20.02.2023/js/main.js
// Algorithm strips all text that follows any of a set of comment markers passed in.
// Any whitespace at the end of the line should also be stripped out.
// Example:

// Given an input string of:
// apples, pears # and bananas
// grapes
// bananas !apples

// The output expected would be:
// apples, pears
// grapes
// bananas
// The code would be called like so:

// var result = solution("apples, pears # and bananas\ngrapes\nbananas !apples", ["#", "!"])
// result should == "apples, pears\ngrapes\nbananas"


/* Solution one */
export const solution = (input, markers) => {
    // Split the input string into lines
    let lines = input.split('\n');

    // Loop through each line
    for (let i = 0; i < lines.length; i++) {
        let line = lines[i];

        // Find the first occurrence of any of the comment markers
        for (let j = 0; j < markers.length; j++) {
            let index = line.indexOf(markers[j]);

            if (index !== -1) {
                // Strip out the comment and any whitespace at the end of the line
                line = line.substring(0, index).trimRight();
                break;
            }
        }

        // Update the line in the array
        lines[i] = line;
    }

    // Join the lines back into a single string and return it
    return lines.join('\n');
}

File: 20.02.2023/py/main.py
# For a given list [x1, x2, x3, ..., xn] compute the last (decimal) digit of x1 ^ (x2 ^ (x3 ^ (... ^ xn))).

# E. g., with the input [3, 4, 2], your code should return 1 because 3 ^ (4 ^ 2) = 3 ^ 16 = 43046721.

# Beware: powers grow incredibly fast. For example, 9 ^ (9 ^ 9) has more than 369 millions of digits. lastDigit has to deal with such numbers efficiently.

# Corner cases: we assume that 0 ^ 0 = 1 and that lastDigit of an empty list equals to 1.


class WrongType(Exception):
    pass

# Solution one
def solution(input_str, markers):
    # Split the input string into lines
    lines = input_str.split('\n')

    # Loop through each line
    for i in range(len(lines)):
        line = lines[i]

        # Find the first occurrence of any of the comment markers
        for marker in markers:
            index = line.find(marker)
            if index != -1:
                # Strip out the comment and any whitespace at the end of the line
                line = line[:index].rstrip()
                break

        # Update the line in the array
        lines[i] = line

    # Join the lines back into a single string and return it
    return '\n'.join(lines)


case = ['apples, pears # and bananas\ngrapes\nbananas !apples', ['#', '!']]

import cProfile
import time
from timeit import default_timer as timer


if __name__ == '__main__':
    cProfile.run('solution(case[0], case[1])')

    start_time = time.time()
    solution(case[0], case[1])
    end_time = time.time()

    print('', end_time - start_time)
